src/main/java/com/celements/cleverreach/DefaultFailNotificationHandler.java

```diff
@@ -0,0 +1,128 @@
+      Optional<String> fromMail = getFromMail(configDoc);
+      if (fromMail.isPresent()) {
+        for (BaseObject receiver : receivers) {
+          if (1 == receiver.getIntValue("is_active")) {
+            mailSender.sendMail(fromMail.get(), null, receiver.getStringValue("email"), null,
+                null, "TAGESAGENDA UPDATE FAILED!", content, content, null, null);
+          }
+        }
+      } else {
+        LOGGER.error("Missing 'from' mail configuration for 'Tagesagenda update failed' "
+            + "notification");
+      }
+    } catch (DocumentNotExistsException dnee) {
+      LOGGER.error("Unable to read failed notification configuration. Doc does not exist.", dnee);
+    }
+  }
+
+  String getMailingContent(String msg, Exception excp) {
+    StringWriter content = new StringWriter();
+    content.append("<h2>")
+        .append(excp.getMessage())
+        .append("</h2><div>")
+        .append(msg).append("</div><hr /><pre>");
+    if ((excp instanceof CleverReachRequestFailedException)
+        && (((CleverReachRequestFailedException) excp).getResponse() != null)) {
+      Response resp = ((CleverReachRequestFailedException) excp).getResponse();
+      content.append("Status Code: ").append(Integer.toString(resp.getStatus())).append("\n");
+      String respHeaders = resp.getStringHeaders().entrySet().stream().collect(
+          Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().stream().collect(Collectors
+              .joining(" | ")))).entrySet().stream().map(entry -> entry.getKey() + " = " + entry
+                  .getValue() + "\n").collect(Collectors.joining());
+      content.append("Header String:\n").append(respHeaders).append("\n");
+      content.append("Body:\n").append(((CleverReachRequestFailedException) excp)
+          .getResponseBody());
+    } else if (excp instanceof CssInlineException) {
+      content.append(((CssInlineException) excp).getExtendedMessage());
+    }
+    appendStackTrace(excp, content);
+    return content.toString();
+  }
+
+  void appendStackTrace(Exception excp, StringWriter content) {
+    StringWriter sw = new StringWriter();
+    try (PrintWriter pw = new PrintWriter(sw)) {
+      excp.printStackTrace(pw);
+      pw.flush();
+      content.append("</pre><hr /><pre>")
+          .append(sw.toString())
+          .append("</pre>");
+    }
+  }
+
+  Optional<String> getFromMail(XWikiDocument configDoc) {
+    return XWikiObjectFetcher.on(configDoc).fetchField(FormMailClass.FIELD_EMAIL_FROM).first()
+        .toJavaUtil();
+  }
+
+  private ClassReference getReceiverEmailClassRef() {
+    return new ClassReference(((OldCoreClasses) oldCoreClasses).getReceiverEmailClassRef(
+        context.getWikiRef().getName()));
+  }
+
+  public DocumentReference getConfigDocRef() {
+    return create(DocumentReference.class, CleverReachRest.REST_CONFIG_DOC_NAME,
+        create(SpaceReference.class, CleverReachRest.REST_CONFIG_SPACE_NAME, context.getWikiRef()));
```

Comment:
in case you don't yet know the newer, more convenient RefBuilder:
```suggestion
    return RefBuilder.from(context.getWikiRef()).space(CleverReachRest.REST_CONFIG_SPACE_NAME)
        .doc(CleverReachRest.REST_CONFIG_DOC_NAME).build(DocumentReference.class);
```